File: Python/chap10/colornames/colornames.py
from PyQt4.QtGui import *
from PyQt4.QtCore import *

class ColorNamesDialog(QDialog):
    def __init__(self, parent=None):
        super(ColorNamesDialog, self).__init__(parent)

        self.sourceModel = QStringListModel(self)
        self.sourceModel.setStringList(QColor.colorNames())

        self.proxyModel = QSortFilterProxyModel(self)
        self.proxyModel.setSourceModel(self.sourceModel)
        self.proxyModel.setFilterKeyColumn(0)

        self.listView = QListView()
        self.listView.setModel(self.proxyModel)
        self.listView.setEditTriggers(QAbstractItemView.NoEditTriggers)

        self.filterLabel = QLabel('&Filter:')
        self.filterLineEdit = QLineEdit()
        self.filterLabel.setBuddy(self.filterLineEdit)

        self.syntaxLabel = QLabel('&Pattern syntax:')
        combo = self.syntaxLComboBox = QComboBox()
        combo.addItem('Regular expression', QRegExp.RegExp)
        combo.addItem('Wildcard', QRegExp.Wildcard)
        combo.addItem('Fixed String', QRegExp.FixedString)
        self.syntaxLabel.setBuddy(combo)

        self.filterLineEdit.textChanged.connect(self.reApplyFilter)
        self.syntaxLComboBox.currentIndexChanged.connect(self.reApplyFilter)

        grid = self.mainLayout = QGridLayout()
        grid.addWidget(self.listView, 0, 0, 1, 2)
        grid.addWidget(self.filterLabel,    1, 0)
        grid.addWidget(self.filterLineEdit, 1, 1)
        grid.addWidget(self.syntaxLabel,    2, 0)
        grid.addWidget(self.syntaxLComboBox,2, 1)
        self.setLayout(grid)

        self.setWindowTitle('Color Names')

    def reApplyFilter(self):
        combo = self.syntaxLComboBox
        index = combo.currentIndex()
        syntax = QRegExp.PatternSyntax(combo.itemData(index))
        regexp = QRegExp(self.filterLineEdit.text(),
                         Qt.CaseInsensitive,
                         syntax)
        self.proxyModel.setFilterRegExp(regexp)

File: Ruby/chap10/colornames/colornames.rb
require 'Qt'

class ColorNamesDialog < Qt::Dialog
  slots 'reApplyFilter()'

  def initialize(parent=nil)
    super(parent)

    @source_model = Qt::StringListModel.new(Qt::Color.colorNames())

    @proxy_model = Qt::SortFilterProxyModel.new(self)
    @proxy_model.setSourceModel(@source_model)
    @proxy_model.setFilterKeyColumn(0)

    @list_view = Qt::ListView.new()
    @list_view.setModel(@proxy_model)
    @list_view.setEditTriggers(Qt::AbstractItemView::NoEditTriggers)

    @filter_label = Qt::Label.new('&Filter:')
    @filter_line_edit = Qt::LineEdit.new()
    @filter_label.buddy = @filter_line_edit

    @syntax_label = Qt::Label.new('&Pattern syntax:')
    @combo = Qt::ComboBox.new()
    @combo.addItem('Regular expression')
    @combo.addItem('Wildcard')
    @combo.addItem('Fixed String')
    @syntax_label.buddy = @combo

    @grid_layout = Qt::GridLayout.new()
    @grid_layout.addWidget(@list_view, 0, 0, 1, 2)

    @grid_layout.addWidget(@filter_label,     1, 0)
    @grid_layout.addWidget(@filter_line_edit, 1, 1)

    @grid_layout.addWidget(@syntax_label, 2, 0)
    @grid_layout.addWidget(@combo,        2, 1)

    setLayout(@grid_layout)
    setWindowTitle('Color Names')

    connect(@filter_line_edit, SIGNAL('textChanged(const QString&)'),
            self, SLOT('reApplyFilter()'))
    connect(@combo, SIGNAL('currentIndexChanged(int)'),
            self, SLOT('reApplyFilter()'))
  end

  def reApplyFilter()
    regexp_table = [
        Qt::RegExp::RegExp,
        Qt::RegExp::Wildcard,
        Qt::RegExp::FixedString
    ]
    index = @combo.currentIndex()
    syntax = regexp_table[index]
    text = @filter_line_edit.text
    regexp = Qt::RegExp.new(text, Qt::CaseInsensitive, syntax)
    @proxy_model.setFilterRegExp(regexp)
  end

end